pkg/controllers/corecluster/corecluster_controller.go

```diff
@@ -0,0 +1,242 @@
+	Infra    *configv1.Infrastructure
+	Platform configv1.PlatformType
+}
+
+// SetupWithManager sets the CoreClusterReconciler controller up with the given manager.
+func (r *CoreClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if err := ctrl.NewControllerManagedBy(mgr).
+		Named(controllerName).
+		For(&configv1.ClusterOperator{}, builder.WithPredicates(clusterOperatorPredicates())).
+		Watches(
+			&clusterv1.Cluster{},
+			handler.EnqueueRequestsFromMapFunc(toClusterOperator),
+			builder.WithPredicates(coreClusterPredicate(r.ManagedNamespace)),
+		).
+		Complete(r); err != nil {
+		return fmt.Errorf("failed to create controller: %w", err)
+	}
+
+	return nil
+}
+
+// Reconcile reconciles the core cluster object for the openshift-cluster-api namespace.
+func (r *CoreClusterReconciler) Reconcile(ctx context.Context, req reconcile.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithName(controllerName)
+	logger.Info("Reconciling core cluster")
+	defer logger.Info("Finished reconciling core cluster")
+
+	ocpInfrastructureName, err := getOCPInfrastructureName(r.Infra)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to obtain ocp infrastructure name: %w", err)
+	}
+
+	cluster, err := r.ensureCoreCluster(ctx, client.ObjectKey{Namespace: r.ManagedNamespace, Name: ocpInfrastructureName}, logger)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to ensure core cluster: %w", err)
+	}
+
+	if !cluster.DeletionTimestamp.IsZero() {
+		if err := r.SetStatusAvailable(ctx, ""); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to set status available: %w", err)
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.ensureCoreClusterControlPlaneInitializedCondition(ctx, cluster); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to ensure core cluster has the ControlPlaneInitializedCondition: %w", err)
+	}
+
+	if err := r.SetStatusAvailable(ctx, ""); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to set status available: %w", err)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// ensureCoreCluster creates a cluster with the given name and returns the cluster object.
+func (r *CoreClusterReconciler) ensureCoreCluster(ctx context.Context, clusterObjectKey client.ObjectKey, logger logr.Logger) (*clusterv1.Cluster, error) {
+	cluster := &clusterv1.Cluster{}
+	if err := r.Client.Get(ctx, clusterObjectKey, cluster); err != nil && !kerrors.IsNotFound(err) {
+		return nil, fmt.Errorf("failed to get core cluster %s/%s: %w", clusterObjectKey.Namespace, clusterObjectKey.Name, err)
+	} else if err == nil {
+		return cluster, nil
+	}
+
+	if r.Infra.Status.PlatformStatus == nil {
+		return nil, errPlatformStatusShouldNotBeNil
+	}
+
+	infraClusterKind, infraClusterAPIVersion, err := mapOCPPlatformToInfraClusterKindAndVersion(r.Platform)
+	if err != nil {
+		return nil, fmt.Errorf("unable to map Infrastucture resource platform type to Cluster API infra cluster kind: %w", err)
```

Comment:
Should be
```suggestion
		return nil, fmt.Errorf("unable to map infrastucture resource platform type to infrastructure cluster kind: %w", err)
```
We are duplicating a lot of the error message here, perhaps we don't need map in both places

Perhaps the inner error should just be `unsupported platform type` and be less detailed?

Reply:
Agree